src/tracking/envelope.cpp

```diff
@@ -125,14 +130,22 @@ namespace impactx
         int num_periods = 1;
         amrex::ParmParse("lattice").queryAddWithParser("periods", num_periods);
 
-        for (int period=0; period < num_periods; ++period)
+        for (period=0; period < num_periods; ++period)
         {
+            // optional, user-defined function call
+            m_tracking_state.m_element = &m_lattice.front();
```

Comment:
Note: We intentionally assign the next element already as we enter the period hook, because it is logical to know the element coming next at this point.
We assign it again in the element loop below.

(Same for the other two tracking loops.)